cpp/log/cert.cc

```diff
@@ -918,21 +902,35 @@ util::Status Cert::IsValidWildcardRedaction() const {
                         "No CT redaction count extension");
   }
 
-  // Unpack the extension contents, which should be SEQUENCE OF INTEGER
-  STACK_OF(ASN1_INTEGER)* const integers(
-      static_cast<STACK_OF(ASN1_INTEGER)*>(
-          ASN1_seq_unpack_ASN1_INTEGER(exty.ValueOrDie()->value->data,
-                                       exty.ValueOrDie()->value->length,
-                                       d2i_ASN1_INTEGER, ASN1_INTEGER_free)));
+  // Ensure the data in the extension is a sequence. DER encoding is same for
+  // SEQUENCE and SEQUENCE OF and we'll check types later.
+  if (exty.ValueOrDie()->value->data[0] !=
+      (V_ASN1_SEQUENCE | V_ASN1_CONSTRUCTED)) {
+    LOG(WARNING) << "CT redaction count extension is not a SEQUENCE OF";
+    return util::Status(Code::INVALID_ARGUMENT,
+                        "CT redaction count extension not a sequence");
+  }
+
+  // Unpack the extension contents, which should be SEQUENCE OF INTEGER.
+  // For compatibility we unpack any sequence and check integer type as we go.
+  // Don't pass the pointer from the extension directly as it gets incremented
+  // during parsing.
+  const unsigned char* sequence_data(
+      const_cast<const unsigned char*>(exty.ValueOrDie()->value->data));
+  STACK_OF(ASN1_TYPE)*
+      const asn1_types(static_cast<STACK_OF(ASN1_TYPE)*>(d2i_ASN1_SEQUENCE_ANY(
+          nullptr,
+          &sequence_data,
+          exty.ValueOrDie()->value->length)));
 
-  if (integers) {
-    const int num_integers = sk_ASN1_INTEGER_num(integers);
+  if (asn1_types) {
+    const int num_integers(sk_ASN1_TYPE_num(asn1_types));
 
     // RFC text says there MUST NOT be more integers than there are DNS ids
     if (num_integers > dns_alt_name_count) {
       LOG(WARNING) << "Too many integers in extension: " << num_integers
                    << " but only " << dns_alt_name_count << " DNS names";
-      FreeOpenSslStackOfAsn1Integer(integers);
+      sk_ASN1_TYPE_pop_free(asn1_types, ASN1_TYPE_free);
```

Comment:
So are you quite sure that sk_FOO_pop_free will free both the container and the contained elements?